kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```diff
@@ -757,87 +760,327 @@ object Resource extends ResourceInstances with ResourcePlatform {
+    Resource applyFull { poll =>
+      poll(fa.allocated).onCancel(fin.use_).map(_.map(fin => (_: Resource.ExitCase) => fin))
+    }
+
+  def uncancelable[A](body: Poll[Resource[F, *]] => Resource[F, A]): Resource[F, A] =
+    Resource applyFull { poll =>
+      val inner = new Poll[Resource[F, *]] {
+        def apply[B](rfb: Resource[F, B]): Resource[F, B] =
+          Resource.applyFull(
+            _(poll(rfb.allocated)).map(_.map(fin => (_: Resource.ExitCase) => fin)))
+      }
+
+      body(inner).allocated.map(_.map(fin => (_: Resource.ExitCase) => fin))
+    }
+}
+
+// note: Spawn alone isn't possible since we need Concurrent to implement start
+abstract private[effect] class ResourceConcurrent[F[_]]
+    extends ResourceMonadCancel[F]
+    with GenConcurrent[Resource[F, *], Throwable] {
+  implicit protected def F: Concurrent[F]
+
+  def never[A]: Resource[F, A] =
+    Resource.eval(F.never[A])
+
+  def cede: Resource[F, Unit] =
+    Resource.eval(F.cede)
+
+  /*
+   * 1. If the scope containing the *fiber* terminates:
+   *   a) if the fiber is incomplete, run its inner finalizers when it complets
+   *   b) if the fiber is succeeded, run its finalizers
+   * 2. If the fiber is canceled or errored, finalize
+   * 3. If the fiber succeeds, await joining and extend finalizers into joining scope
+   * 4. If a fiber is canceled *after* it completes, we explicitly snag the finalizers and run them, setting the result to Canceled()
+   *   a) Note that we are missing a check here! We need to disable this behavior if someone has already joined
+   */
+  def start[A](fa: Resource[F, A]): Resource[F, Fiber[Resource[F, *], Throwable, A]] = {
+    final case class State(
+        fin: F[Unit] = F.unit,
+        runOnComplete: Boolean = false,
+        completed: Boolean = false,
+        canceled: Boolean = false)
+
+    Resource {
+      import Outcome._
+
+      F.ref[State](State()) flatMap { state =>
+        val folded =
+          fa.fold(_.pure[F], release => state.update(s => s.copy(fin = s.fin !> release)))
+
+        val finalized = F.guaranteeCase(folded) {
+          case Canceled() | Errored(_) =>
+            state.get.flatMap(_.fin)
+
+          case Succeeded(_) =>
+            val maybeFins = state modify { s =>
+              (
+                s.copy(completed = true),
+                Some(s.fin).filter(_ => s.runOnComplete && !s.canceled))
+            }
+
+            maybeFins.flatMap(_.getOrElse(F.unit))
+        }
+
+        F.start(finalized) map { outer =>
+          val fiber = new Fiber[Resource[F, *], Throwable, A] {
+
+            def cancel =
+              Resource eval {
+                F uncancelable { poll =>
+                  // technically cancel is uncancelable, but separation of concerns and what not
+                  val maybeFins = poll(outer.cancel) *> {
+                    state modify { s =>
+                      (s.copy(canceled = true), Some(s.fin).filter(_ => s.completed))
```

Comment:
```suggestion
                      if (!s.joined)
                        (s.copy(canceled = true, fin = None), s.fin.traverse_(x => x))
                      else
                        (s.copy(canceled = true), F.unit)
```
How I think it'd look with `joined`.